livekit-agents/livekit/agents/utils/connection_pool.py

```diff
@@ -0,0 +1,166 @@
+        self._to_close: Set[T] = set()
+
+    async def _connect(self) -> T:
+        """Create a new connection.
+
+        Returns:
+            The new connection object
+
+        Raises:
+            NotImplementedError: If no connect callback was provided
+        """
+        if self._connect_cb is None:
+            raise NotImplementedError("Must provide connect_cb or implement connect()")
+        return await self._connect_cb()
+
+    async def _drain_to_close(self) -> None:
+        """Drain and close all the connections queued for closing."""
+        for conn in list(self._to_close):
+            await self._maybe_close_connection(conn)
+        self._to_close.clear()
+
+    @asynccontextmanager
+    async def connection(self) -> AsyncGenerator[T, None]:
+        """Get a connection from the pool and automatically return it when done.
+
+        Yields:
+            An active connection object
+        """
+        conn = await self.get()
+        try:
+            yield conn
+        except Exception:
+            self.remove(conn)
+            raise
+        else:
+            self.put(conn)
+
+    async def get(self) -> T:
+        """Get an available connection or create a new one if needed.
+
+        Returns:
+            An active connection object
+        """
+        await self._drain_to_close()
+
+        now = time.time()
+
+        # try to reuse an available connection that hasn't expired
+        while self._available:
+            conn = self._available.pop()
+            if (
+                self._max_session_duration is None
+                or now - self._connections[conn] <= self._max_session_duration
+            ):
+                return conn
+            # connection expired; mark it for resetting.
+            self.remove(conn)
+
+        conn = await self._connect()
+        self._connections[conn] = now
+        return conn
+
+    def put(self, conn: T) -> None:
+        """Mark a connection as available for reuse.
+
+        If connection has been reset, it will not be added to the pool.
+
+        Args:
+            conn: The connection to make available
+        """
+        if conn in self._connections:
+            self._available.add(conn)
+
+    async def _maybe_close_connection(self, conn: T) -> None:
+        """Close a connection if close_cb is provided.
+
+        Args:
+            conn: The connection to close
+        """
+        if self._close_cb is not None:
+            await self._close_cb(conn)
+
+    def remove(self, conn: T) -> None:
+        """Remove a specific connection from the pool.
+
+        Marks the connection to be closed during the next drain cycle.
+
+        Args:
+            conn: The connection to reset
+        """
+        self._available.discard(conn)
+        if conn in self._connections:
+            self._to_close.add(conn)
+            self._connections.pop(conn, None)
+
+    def maybe_remove(self, conn: T) -> None:
```

Comment:
I'm not a fan of this "maybe_remove". Hard to grasp what it is doing initially. Would be great if you don't have to think about it

Reply:
right. it's basically opportunistically handling removing the connection at the right moment. def a bit harder to understand

Reply:
perhaps we don't need this.. and relying on the timeout is sufficient..

Reply:
yes, relying on timeout should be sufficient IMO. I see only Google uses it once again, and double checking the code, the Google code could even just use the `remove` function since it is already checking for expiration:
```python
                        if (
                            time.time() - self._session_connected_at
                            > _max_session_duration
                        ):
                            logger.debug(
                                "Google STT maximum connection time reached. Reconnecting..."
                            )
                            self._pool.maybe_remove(client)
```